hackett-lib/hackett/private/adt.rkt

```diff
@@ -256,7 +264,151 @@
+                     (ideal->subst-fn q x)))
+       (λ (r)
+         (ideal-con ctor
+                    (for/list ([fn (in-list fns)])
+                      (fn r))))]
+      [_ (λ (r) q)]))
+
+  ;; Substitute occurences of symbol 'x' with each ideal in 'rs',
+  ;; for each ideal in 'qs'.
+  ;; e.g.
+  ;;   (subs 'A '(B C) (list 'D (con "*" 'A)))
+  ;;   =
+  ;;   (list (list 'D (con "*" 'B))
+  ;;         (list 'D (con "*" 'C)))
+  (define (substitute-ideals x rs qs)
+    (define subst-fns (for/list ([q (in-list qs)])
+                        (ideal->subst-fn q x)))
+    (for/list ([r (in-list rs)])
+      (for/list ([fn (in-list subst-fns)])
+        (fn r))))
+
+  (define current-exhaust-split-handler
+    (make-parameter #f))
+
+  ;; Returns #t if the ideals satisfy the patterns.
+  ;; Calls (current-exhaust-split-handler) when an ideal variable
+  ;; needs to be split.
+  (define (ideals-satisfied? ideals pats)
+    (for/and ([p (in-list pats)]
+              [q (in-list ideals)])
+      (match p
+        ; The ideal is satisfied when we hit a wildcard pattern, such as
+        ; a variable or a hole
+        [(pat-var _ _) #t]
+        [(pat-hole _) #t]
+
+        ; When we hit a constructor pattern, we check the ideal. If it is
+        ; a constructor, compare the tags and then recur for the sub-patterns.
+        ; If it is a variable, then split the ideal into new ideals for each kind
+        ; of constructor.
+        [(pat-con _ ctor sub-pats)
+         (match q
+           [(ideal-con ctor-tag sub-ideals)
+            (and (eq? (syntax-local-value ctor-tag) ctor)
+                 (ideals-satisfied? sub-ideals sub-pats))]
+
+           [(? symbol? x)
+            (define split-into
+              (for/list ([ctor-tag (in-list (data-constructor-all-tags ctor))])
+                (constructor-tag->ideal-con ctor-tag)))
+            ((current-exhaust-split-handler) x split-into)])]
+
+        ; TODO: better exhaustiveness checking on strings.
+        ; Ocaml checks for the strings "*", "**", "***" etc. It would
+        ; be fairly easy to do the same using splitting.
+        [(pat-str _ s) #f])))
+
+
+  ;; Exhaustiveness checking algorithm. Given a list of pattern-lists,
+  ;; returns #f if no un-matched patterns found. Otherwise,
+  ;; returns an example of an un-matched pattern-list.
+  (define/contract (exhaustive patss pat-count)
+    (-> (listof (listof pat?))
+        exact-nonnegative-integer?
+        (or/c #f (listof ideal?)))
+
+    (define check
+      (match-lambda
+        ; No more ideals to check; #f signals that the
+        ; pattern is exhaustive.
+        ['() #f]
+
+        ; Check if the most recent ideal is exhaustive, or if it split
+        ; into more ideals.
+        [(cons ideals rest-idealss)
+         (define sat
+           (let/ec escape ; This continuation is used to escape in case of a split
+             (parameterize ([current-exhaust-split-handler
```

Comment:
This seems a little unnecessary to me. My understanding is that continuations are relatively expensive in Racket (even escape continuations), so I would generally like to avoid capturing them in a loop. *Performance aside*, the non-local control flow here seems like it would be extremely easy to eliminate, since `ideals-satisfied?` could just produce a different sort of value. Do you think it would be unreasonable to replace this parameter with a simpler signaling mechanism?

Reply:
I thought it was a fine place to use non-local control flow, since breaking out of recursion is generally a good place to use continuations. However, I don't know much about the performance characteristics.

You certainly could use a special value instead, maybe taking advantage of falsity (for/or, for/and). It might make the function a bit more complicated to describe, but if you think it is worth getting rid of the `let/ec` then you should go ahead.